packages/dataviews/src/components/dataviews/index.tsx

```diff
@@ -53,29 +58,42 @@ type DataViewsProps< Item > = {
 	? { getItemId?: ( item: Item ) => string }
 	: { getItemId: ( item: Item ) => string } );
 
+export type DataViewsCompositionProps< Item > = {
+	children: ReactNode;
+	data?: Item[];
+} & Partial< Omit< DataViewsControlledProps< Item >, 'children' | 'data' > >;
```

Comment:
Any specific reason for not just adding `children?: ReactNode` to `DataViewsProps` and keeping the rest the same?
I understand that maybe not all these items are necessary anymore as a consumer may not use all aspects of the nested DataViews structure, but if they do, they still require all the above props to be passed in.

IMO, making everything else optional when `children` is passed in convolutes the types. I would suggest just adding `children?: ReactNode`.

Reply:
Same here. I think everything else should not be optional actually. There's just the "header" prop that don't make much sense in the context of the "provider". All the rest seems like it should remain as is.

Reply:
I’m sorry, but this isn’t how TypeScript is supposed to work. Its primary purpose is to help us anticipate problems during static analysis, *before* we even run the code.

If we make all props optional, what’s the point of having types? We’ll be able to do something like:

TypeScript will be happy, but then at runtime…

```
normalize-fields.ts:33 Uncaught TypeError: Cannot read properties of undefined (reading 'map')
    at normalizeFields (normalize-fields.ts:33:1)
    at index.tsx:104:1
    at DataViews (index.tsx:104:1)
```

Clearly, we’re not using TypeScript as intended. That leads me to the question: if we allow this, why use TypeScript at all instead of plain JavaScript?

Reply:
I think what we're saying is not to make everything optional, but to just make `children` optional because that's our intention here. 

I think what @louwie17 is saying is that you made everything optional by using `Partial`

Reply:
Ah, got it, pardon!. thanks for clarifying!

But in that case, we’re just hitting the other side of the same coin: if we don’t make the props optional, then TypeScript will require all of them, even in the free composition mode, where many of them might not make sense.

For the same example above, it will throw tons of TypeScript errors, even though this is a valid and intentional use case.

That’s actually the reason we introduced the two separate types, to reflect the two modes DataViews supports: the controlled mode (the current one), and the free composition mode (using children).

The idea is to align the typing with how the component is being used, based on whether children is present or not.

Reply:
>The idea is to align the typing with how the component is being used, based on whether children is present or not. 

Yes, it seems you consider the props of the provider different than the props of the regular DataViews. This seems to be the main difference here. Why do you think so, why should these props be optional for the provider but mandatory for DataViews?

Reply:
> Why do you think so, why should these props be optional for the provider but mandatory for DataViews?

Because free composition changes the rules around which props are required.

For example, `children` and `data` are clearly required, and there are likely other props that should also be required in this mode, but that’s something we still need to analyze carefully.

At the same time, depending on the composition, some props might not be needed. For instance, if we’re not rendering a `<Pagination />` component, do we really need to define the `paginationInfo` prop?

Reply:
@jsnajdr I think this Automattic/wp-calypso#102987, which is aligned with your suggestion. I took that approach precisely to address the `onClickItem` issue.

The name is `<DataViewsProvider />`,

Reply:
> At the same time, depending on the composition, some props might not be needed. For instance, if we’re not rendering a <Pagination /> component, do we really need to define the paginationInfo prop?

For me paginationInfo has more to do about the "data" and not about whether pagination is rendered or not. In other words, I don't see this as something to differentiate the props between controlled UI and uncontrolled UI. I can be onboard with making this prop optional but I would do so for all use-cases not just the "uncontrolled UI".

Reply:
Okay, sounds good to me. Let’s iterate on it then